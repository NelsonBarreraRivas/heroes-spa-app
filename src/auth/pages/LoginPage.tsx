import { FormEvent, useContext } from "react"
import { useNavigate } from "react-router-dom"
import { LoadingPage } from "../../ui/components/LoadingPage"
import { AuthContext } from "../context"


export const LoginPage = () => {

    const { onLogin } = useContext( AuthContext )

    const navigate = useNavigate()

    const handleSubmit = (event : FormEvent)  => {

        event.preventDefault()

        const lastPath = localStorage.getItem('lastPath') ?? '/'

        onLogin('Nelson', 'nelsonbr')

        navigate(lastPath, {replace: true} )
    }


    return (
        <div className="relative">

            <LoadingPage time={500} />

            <div className="flex min-h-full items-center justify-center py-12 px-4 sm:px-6 lg:px-8">

                <div className="w-full max-w-md space-y-8">
                    <div>
                        <h2 className="mt-6 text-center text-5xl font-bold tracking-tight text-gray-900">Sign in to your account</h2>

                    </div>
                    <form className="mt-8 space-y-6" onSubmit={ handleSubmit }>
                        <input type="hidden" name="remember" value="true" />
                        <div className="flex flex-col gap-4 rounded-md shadow-sm">
                            <div>
                                <label className="sr-only">Name</label>
                                <input name="text" type="name" className="outline-none block w-full rounded-t-md border-0 py-3 text-base px-5 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-red-600 sm:leading-6" placeholder="Username" />
                            </div>
                            <div>
                                <label className="sr-only">Email address</label>
                                <input name="email" type="email" className="outline-none block w-full rounded-t-md border-0 py-3 text-base px-5 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-red-600 sm:leading-6" placeholder="Email address" />
                            </div>
                            <div>
                                <label className="sr-only">Password</label>
                                <input  name="password" type="password" className="outline-none block w-full rounded-b-md border-0 py-3 text-base px-5 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-red-600 sm:leading-6" placeholder="Password" />
                            </div>
                        </div>



                        <div>
                            <button className="flex w-full justify-center rounded-md bg-red-600 py-2 px-3 text-lg font-semibold text-white hover:bg-red-500">
                                Sign in
                            </button>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}
