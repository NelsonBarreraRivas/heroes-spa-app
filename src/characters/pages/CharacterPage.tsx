import { List } from "../../ui/components/List"
import { ListSkeleton } from "../../ui/components/skeleton/ListSkeleton"
import useCharacter from "../hooks/useCharacter"


export const CharacterPage = () => {

    const { data, error, loading } = useCharacter()


    return (
        <>
            <h1 className="text-center text-4xl pt-6 underline">Characters Page</h1>
            {
                loading ?
                    <ListSkeleton numberSkeletons={4} typeList={{ type : 'multiple-cols'}}/> :
                    <List data={ data || [] } typeList={{ type : 'multiple-cols'}}/>
            }
        </>
    )
}
