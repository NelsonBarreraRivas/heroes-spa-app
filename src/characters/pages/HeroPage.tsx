import { useState } from "react";
import { Link, Navigate, useNavigate, useParams } from "react-router-dom"
import useCharacterId from "../hooks/useCharacterId";

export const HeroPage = () => {

    const { id = ''} = useParams()

    const navigate = useNavigate()

    const { loading, data } = useCharacterId({ characterId: Number(id) })

    const onNavigateBack = () => {
        navigate(-1);
    }


    if (!loading && !data) {
        return <Navigate to={'/error-404'} replace />
    }


    return (
        !loading && data ?
            (
                <div className="my-8">
                    <div className="flex flex-col md:flex-row items-center justify-center gap-6">
                        <h1 className="text-4xl text-center">{data?.name}</h1>
                        <button onClick={onNavigateBack} className="block justify-center items-center px-3 py-2 text-base font-medium text-center text-white bg-red-600 rounded-lg hover:bg-red-700 focus:ring-4 focus:outline-none">
                            Go Back
                        </button>
                    </div>


                    <article className="grid grid-cols-1 md:grid-cols-2 gap-6 mt-8" aria-label="article-character">
                        <div className="animate__animated animate__fadeInLeft">
                            <img className="object-cover" src={data?.thumbnail.path + '.' + data?.thumbnail.extension} alt={data?.name} />

                        </div>
                        <div>
                            <p className="text-2xl text-red-600">Comics:</p>
                            <ul>
                                {data?.comics.items.map(comic => <li key={comic.name}>{comic.name}</li>)}
                            </ul>
                            <p className="text-2xl text-red-600">Stories:</p>
                            <ul>
                                {data?.stories.items.map(story => <li key={story.name}>{story.name}</li>)}
                            </ul>
                            <p className="text-2xl text-red-600">Series:</p>
                            <ul>
                                {data?.series.items.map(serie => <li key={serie.name}>{serie.name}</li>)}
                            </ul>
                        </div>
                    </article>
                </div>
            ) :
            (
                <></>
            )
    )
}
