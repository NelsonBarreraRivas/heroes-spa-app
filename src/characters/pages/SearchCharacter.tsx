import queryString from "query-string"
import { FormEvent } from "react"
import { useLocation, useNavigate } from "react-router-dom"
import { useForm } from "../../hooks/useForm"
import { List } from "../../ui/components/List"
import { ListSkeleton } from "../../ui/components/skeleton/ListSkeleton"
import useCharacterByName from "../hooks/useCharacterByName"


export const SearchCharacter = () => {

    const navigate = useNavigate()
    
    const { search } = useLocation()

    const { q = '' } = queryString.parse(search)

    const query = (q && typeof q === "string") ? q.toLowerCase().trim() : ''

    const { data, loading } = useCharacterByName({ nameStartsWith: q })


    const showSearch = (q?.length === 0)
    const showError = (q!.length > 0) && data?.length === 0

    const { onChangue, onResetForm, searchText } = useForm({
        searchText: query
    })

    const onSubmit = (e: FormEvent) => {
        e.preventDefault()
        navigate(`?q=${searchText}`)
        onResetForm()
    }


    return (
        <div className="">
            <h1 className="text-center text-4xl py-6 underline">Search Character</h1>
            <div className="grid grid-cols-1 sm:grid-cols-12 gap-8">
                <div className="sm:col-span-5">
                    <h4 className="text-2xl">Searching:</h4>
                    <form onSubmit={onSubmit} role='form'>
                        <input
                            name="searchText"
                            type="search"
                            value={searchText}
                            onChange={onChangue}
                            className="outline-none block w-full rounded-t-md border-0 py-3 text-base px-5 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:z-10 focus:ring-2 focus:ring-inset focus:ring-red-600 sm:leading-6"
                            placeholder="Search.... Enter."
                            autoComplete="off"
                        />

                    </form>
                </div>
                <div className="sm:col-span-7">
                    <h4 className="text-2xl">Results:</h4>
                    <div style={{ display: showSearch ? '' : 'none' }} className="animate__animated animate__fadeIn bg-blue-100 border-t border-b border-blue-500 text-blue-700 px-4 py-3" aria-label="alert">
                        <p className="font-bold">Informational message</p>
                        <p className="text-base">Search to Character.</p>
                    </div>
                    <div style={{ display: showError ? '' : 'none' }} className="animate__animated animate__fadeIn bg-red-100 border-t border-b border-red-500 text-red-500 px-4 py-3" aria-label="alert">
                        <p className="font-bold">Informational message</p>
                        <p className="text-base">There's no result with <b>{q}</b>.</p>
                    </div>

                    {
                        loading ?
                            <ListSkeleton numberSkeletons={1} typeList={{ type: 'simple-cols' }} /> :
                            <List data={data || []} typeList={{ type: 'simple-cols' }} />

                    }
                </div>
            </div>
        </div>
    )
}
