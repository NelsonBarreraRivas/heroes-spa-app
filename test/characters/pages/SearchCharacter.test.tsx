import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React, { ChangeEvent } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { SearchCharacter } from '../../../src/characters/pages/SearchCharacter'

const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate : () => mockedUseNavigate
}))

describe('Pruebas en <SearchCharacter />', () => {

    const character = 'hulk'

    test('debe de mostrarse correctamente con los valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchCharacter />
            </MemoryRouter>
        )

        expect( container ).toMatchSnapshot()
    });

    test('debe de mostrar a Hulk y el input con el valor del queryString', async () => {

        render(
            <MemoryRouter initialEntries={[`/search-character?q=${ character }`]}>
                <SearchCharacter />
            </MemoryRouter>
        )

        await waitFor(() => screen.getByTestId('list-data'))

        const inputElement = screen.getByRole('searchbox') as HTMLInputElement

        const alertsRole = screen.getAllByLabelText('alert') 

        alertsRole.forEach( alert => expect(alert.style.display).toBe('none') )

        expect( inputElement.value ).toBe( character )
    
        expect( screen.getByTestId('list-data') ).toBeTruthy()

    });

    test('debe de mostrar un error si no se encuentra el charcter (batman) ', async () => {
        
        const characterNone = 'batman'

        render(
            <MemoryRouter initialEntries={[`/search-character?q=${ characterNone }`]}>
                <SearchCharacter />
            </MemoryRouter>
        )

        await waitFor(() => screen.getByTestId('list-data'))

        const list = screen.getByTestId('list-data')

        expect(list.querySelector('li')).toBeFalsy();

        const alertsRole = screen.getAllByLabelText('alert') 

        
        const [ alert ] = alertsRole.filter( alert => alert.style.display !== 'none')

        expect( alert.textContent ).toContain(`no result with ${ characterNone }`)

    });

    test('debe de llamar al navigate a la pantalla nueva cuando se dispare el submit', async () => {
        
        render(
            <MemoryRouter initialEntries={[`/search-character`]}>
                <SearchCharacter />
            </MemoryRouter>
        )

        const inputElement = screen.getByRole('searchbox') as HTMLInputElement

        fireEvent.change( inputElement, { target: { value: character } } as ChangeEvent<HTMLInputElement> )

        const formElement = screen.getByRole('form')

        fireEvent.submit( formElement )

        expect( mockedUseNavigate ).toBeCalledWith(`?q=${character}`)

    });

});