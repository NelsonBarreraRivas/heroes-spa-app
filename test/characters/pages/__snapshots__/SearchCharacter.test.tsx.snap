// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchCharacter /> debe de mostrarse correctamente con los valores por defecto 1`] = `
<div>
  <div
    class=""
  >
    <h1
      class="text-center text-4xl py-6 underline"
    >
      Search Character
    </h1>
    <div
      class="grid grid-cols-1 sm:grid-cols-12 gap-8"
    >
      <div
        class="sm:col-span-5"
      >
        <h4
          class="text-2xl"
        >
          Searching:
        </h4>
        <form
          role="form"
        >
          <input
            autocomplete="off"
            class="outline-none block w-full rounded-t-md border-0 py-3 text-base px-5 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:z-10 focus:ring-2 focus:ring-inset focus:ring-red-600 sm:leading-6"
            name="searchText"
            placeholder="Search.... Enter."
            type="search"
            value=""
          />
        </form>
      </div>
      <div
        class="sm:col-span-7"
      >
        <h4
          class="text-2xl"
        >
          Results:
        </h4>
        <div
          aria-label="alert"
          class="animate__animated animate__fadeIn bg-blue-100 border-t border-b border-blue-500 text-blue-700 px-4 py-3"
        >
          <p
            class="font-bold"
          >
            Informational message
          </p>
          <p
            class="text-base"
          >
            Search to Character.
          </p>
        </div>
        <div
          aria-label="alert"
          class="animate__animated animate__fadeIn bg-red-100 border-t border-b border-red-500 text-red-500 px-4 py-3"
          style="display: none;"
        >
          <p
            class="font-bold"
          >
            Informational message
          </p>
          <p
            class="text-base"
          >
            There's no result with 
            <b />
            .
          </p>
        </div>
        <ul
          class="grid grid-cols-1 gap-4 my-8 justify-center justify-items-center"
          data-testid="list-data"
        />
      </div>
    </div>
  </div>
</div>
`;
